dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,190 @@
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                TOP_10_PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID, COUNT(DISTINCT USER_ID) AS N_UNIQUE_USERS
+                    FROM RECENT_DOWNLOADS
+                    GROUP BY PROJECT_ID
+                    ORDER BY N_UNIQUE_USERS DESC
+                    LIMIT 10
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM TOP_10_PUBLIC_PROJECTS)
+                ),
+                FILE_SIZES AS (
+                    SELECT LATEST_FILE_HANDLES.PROJECT_ID, SUM(FILE_LATEST.CONTENT_SIZE) / POWER(1024, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                    FROM LATEST_FILE_HANDLES
+                    JOIN SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILE_LATEST
+                    ON LATEST_FILE_HANDLES.FILE_HANDLE_ID = FILE_LATEST.ID
+                    GROUP BY LATEST_FILE_HANDLES.PROJECT_ID
+                )
+
+                SELECT 
+                    TOP_10_PUBLIC_PROJECTS.PROJECT_ID,
+                    TOP_10_PUBLIC_PROJECTS.N_UNIQUE_USERS,
+                    COALESCE(TO_CHAR(MAX(RECENT_DOWNLOADS.RECORD_DATE), 'YYYY-MM-DD'), 'N/A') AS LAST_DOWNLOAD_DATE,
+                    ROUND(FILE_SIZES.TOTAL_DATA_SIZE_IN_GIB, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                FROM 
+                    TOP_10_PUBLIC_PROJECTS
+
+                LEFT JOIN RECENT_DOWNLOADS
+                ON TOP_10_PUBLIC_PROJECTS.PROJECT_ID = RECENT_DOWNLOADS.PROJECT_ID
+
+                JOIN FILE_SIZES
+                ON TOP_10_PUBLIC_PROJECTS.PROJECT_ID = FILE_SIZES.PROJECT_ID
+
+                GROUP BY
+                    TOP_10_PUBLIC_PROJECTS.PROJECT_ID,
+                    TOP_10_PUBLIC_PROJECTS.N_UNIQUE_USERS,
+                    FILE_SIZES.TOTAL_DATA_SIZE_IN_GIB
+
+                ORDER BY 
+                    TOP_10_PUBLIC_PROJECTS.N_UNIQUE_USERS DESC;
+            """
+
+        cs.execute(query)
+        trending_project_snapshot = cs.fetch_pandas_all()
+
+        metrics = []
+        for _, row in trending_project_snapshot.iterrows():
+            metrics.append(
+                SnapshotMetrics(
+                    project_id=row["PROJECT_ID"],
+                    n_unique_users=row["N_UNIQUE_USERS"],
+                    last_download_date=row["LAST_DOWNLOAD_DATE"],
+                    total_data_size_in_gib=row["TOTAL_DATA_SIZE_IN_GIB"],
+                )
+            )
+        return metrics
+
+    @task
+    def push_results_to_synapse_table(metrics: List[SnapshotMetrics], **context) -> None:
+        """Push the results to a Synapse table."""
+        data = []
+        export_date = date.today()
```

Comment:
This `export_date` may lead to some issues if we are backfilling data. I am thinking of the situation where "I want to load the data from Jan 2024". In that case it is correct that I would be exporting the data today, however, the data being exported represents data for the month of Jan 2024.

Maybe this would need to change to `month_to_run` instead of `date.today()`?

Reply:
Good point. `LAST_DOWNLOAD_DATE` shares the date of the last download within the designated month for the query, so technically this is the column that shares which month the data pertains to: If you take a look at my [Trending Snapshots Test](https://www.synapse.org/Synapse:syn61932294/tables/) table you'll see that I exported data from both June and July on the same day, as denoted by `LAST_DOWNLOAD_DATE` and `EXPORT_DATE`.

One solution we can go with is renaming `LAST_DOWNLOAD_DATE` to something more specific like `LAST_DOWNLOAD_OF_MONTH`. What do you think of moving forward this way?

Reply:
`LAST_DOWNLOAD_DATE` should be sufficient for what you need here. So then I think you'd be able to query for a month worth of data like:


`SELECT * FROM syn61932294 where LAST_DOWNLOAD_DATE >= '1717200000000' AND  LAST_DOWNLOAD_DATE < '1719792000000'`

1717200000000 = `Sat Jun 01 2024 00:00:00.000` UTC
1719792000000 = `Mon Jul 01 2024 00:00:00.000` UTC

Reply:
The way we are querying for data is with the `DATE_TRUNC` operator on Snowsight's SQL. So it retrieves downloads for a given month based on the month_to_download given in the dag params:

<img width="546" alt="image" src="https://github.com/user-attachments/assets/88f6662b-bba9-40d3-9b99-3e21be4d51b8">

Reply:
Right, that's on the snowflake side. 

The query I gave would be what you might use to select data from the Synapse table

Reply:
I see what you mean. Yes, that's right! 
<img width="869" alt="image" src="https://github.com/user-attachments/assets/76cf3a76-91f4-48ee-98ff-67f94edfa2af">